cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,145 @@
+  Unknown,
+  // Target is a regular file
+  File,
+  // Target is a directory
+  Directory
+};
+
+std::string ToString(FileType);
+
+struct FileStats {
+  FileType type = FileType::Unknown;
+  // The file path in the filesystem.
+  std::string path;
+  // The size in bytes, if available.  Only regular files are guaranteed
+  // to have a size.
+  int64_t size = -1;
+  // The time of last modification, if available.
+  TimePoint mtime = TimePoint(-1);
+
+  FileStats() = default;
+
+  // The file base name (component after the last '/').
+  std::string base_name() const;
+};
+
+struct OpenFlags {
+  // Whether to transparently compress or decompress, based on file extension
+  // or other heuristics.
+  bool transparent_compression = true;
+
+  OpenFlags() = default;
+};
+
+struct Selector {
+  // The folder in which to select files.
+  std::string base_folder;
+  // Whether to recurse into subfolders.
+  bool recursive = false;
+
+  MatchOptions() = default;
+};
+
+
+class FileSystem {
+ public:
+  virtual ~FileSystem();
+
+  /// Get statistics for the given target.
+  ///
+  /// Any symlink is automatically dereferenced, recursively
+  virtual Status GetTargetStats(const std::string& path, FileStats* out);
+  /// Same, for many targets at once.
+  virtual Status GetTargetStats(const std::vector<std::string>& paths,
+                                std::vector<FileStats>* out) = 0;
+  virtual Status GetTargetStats(const Selector& select,
+                                std::vector<FileStats>* out) = 0;
+
+  /// Create a folder and subfolders.
+  virtual Status CreateFolder(const std::string& path, bool recursive = true) = 0;
+
+  /// Delete a folder and contents.
+  virtual Status DeleteFolder(const std::string& path) = 0;
+
+  /// Delete a file.
+  virtual Status DeleteFile(const std::string& path) = 0;
+  /// Delete many files.
+  virtual Status DeleteFiles(const std::vector<string>& path) = 0;
+
+  /// Move / rename a file or directory.
+  ///
+  /// The destination will be replaced if it exists.
+  virtual Status Move(const std::string& src, const std::string& dest) = 0;
+
+  /// Copy a file.
+  ///
+  /// The destination will be replaced if it exists.
+  virtual Status Copy(const std::string& src, const std::string& dest) = 0;
+
+  /// Open an input stream.
+  virtual Status OpenInputStream(const std::string& path,
+                                 std::shared_ptr<InputStream>* out);
+
+  virtual Status OpenInputStream(const std::string& path, OpenFlags flags,
+                                 std::shared_ptr<InputStream>* out) = 0;
+
+  /// Open an output stream.
+  ///
+  /// If the target already exists, it is replaced with a new file.
+  virtual Status OpenOutputStream(const std::string& path,
+                                  std::shared_ptr<OutputStream>* out);
+
+  virtual Status OpenOutputStream(const std::string& path, OpenFlags flags,
+                                  std::shared_ptr<OutputStream>* out) = 0;
```

Comment:
Should consider APIs for appending to existing files. Here is what TF does

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/file_system.h#L78

Reply:
What is appending useful for in our context?

Reply:
? You would want append if you wanted to make a generally useful file system abstraction, I would think.

Reply:
Right, but here the context is the Arrow datasets project.

Reply:
I am in this conversation for somewhat of a broader context...
I hope that arrow is trying to make a filesystem interface that is useful not just for the internal readers and otherwise unavailable or feature-poor for the other possible users across several languages.

Reply:
@pitrou you could append to an IPC stream in a file, or a Parquet `_metadata` file, for example.

Reply:
Ok, add some appending APIs.